main/commons/src/main/java/org/cryptomator/common/mountpoint/MountPointChooser.java

```diff
@@ -0,0 +1,170 @@
+package org.cryptomator.common.mountpoint;
+
+import dagger.MapKey;
+import org.cryptomator.common.vaults.Volume;
+
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Base interface for the Mountpoint-Choosing-Operation that results in the choice and
+ * preparation of a Mountpoint or an exception otherwise.<br>
+ * <p>All <i>MountPointChoosers (MPCs)</i> need to implement this class and must be added to
+ * the pool of possible MPCs by {@link MountPointChooserModule MountPointChooserModule.}
+ * The MountPointChooserModule requires all {@link dagger.Provides Providermethods} to
+ * be annotated with the {@link PhaseKey @PhaseKey-Annotation} and a <b>unique</b> Phase to
+ * allow the Module to sort them according to the phases' {@link Phase#getTiming()} timing.}
+ * The timing must be defined by the developer to reflect a useful execution order.
+ *
+ * <p><b>Phase-Uniqueness:</b> Phases must be unique, meaning that they must not be used
+ * to annotate more than <i>one</i> Providermethod. Define a new phase for each additional
+ * MPC that is added to the Module. Timings can be reused; the order of Phases with equal
+ * timing among themselves is undefined.
+ *
+ * <p>MPCs are executed by a {@link Volume} in the order of their phase's timing to find
+ * and prepare a suitable Mountpoint for the Volume. The Volume only has access to a {@link Set}
+ * of MPCs in this specific order, that is provided by the Module; the according Phases and exact
+ * timings are inaccessible to the Volume. The Set only contains Choosers that were deemed
+ * {@link #isApplicable() applicable} by the Module.
+ *
+ * <p>At execution of a MPC {@link #chooseMountPoint()} and then {@link #prepare(Path)} are called
+ * by the Volume. If {@code #chooseMountPoint()} yields no result, the next MPC is executed
+ * without first calling the {@code #prepare(Path)}-Method of the current MPC.
+ * This is repeated until<br>
+ * <ul>
+ *     <li><b>either</b> a Mountpoint is returned by {@code #chooseMountPoint()}
+ *     and {@code #prepare(Path)} succeeds or fails</li>
+ *     <li><b>or</b> no MPC remains and an {@link InvalidMountPointException} is thrown.</li>
+ * </ul>
+ * If the {@code #prepare(Path)}-Method of a MPC fails the entire Mountpoint-Choosing-Operation
+ * is aborted and the method should do all necessary cleanup before throwing the exception.
+ * If the preparation succeeds {@link #cleanup(Path)} can be used after unmount to do any
+ * remaining cleanup.
```

Comment:
Change this! If the chosen method failed, show an error. DO NOT proceed with a different method. The user needs to decide for a different mount option. It'd be confusing, if one configures a custom mount path, but the vault is instead mounted to a drive letter and reports "success".

There are users who automate access to vaults and might copy data onto the alleged mount point, if mounting succeeded. This can even cause leakage of sensitive data.

Reply:
Resolved in personal conversation.